goyaml.v3/emitterc.go

```diff
@@ -1144,8 +1149,11 @@ func yaml_emitter_process_head_comment(emitter *yaml_emitter_t) bool {
 }
 
 // Write an line comment.
-func yaml_emitter_process_line_comment(emitter *yaml_emitter_t) bool {
+func yaml_emitter_process_line_comment(emitter *yaml_emitter_t, linebreak bool) bool {
 	if len(emitter.line_comment) == 0 {
+		if linebreak && !put_break(emitter) {
+			return false
+		}
```

Comment:
what is this changing from yaml.v3 default behavior? is it configurable?

Reply:
We have an issue that describes in detail what's being changed https://github.com/kubernetes-sigs/kustomize/issues/4033

We aren't changing anything about yaml.v3 default behavior, but we are adding an option to configure the style of sequence indents.

In older versions of yaml v3, we had:
```
foo:
- bar
```

Then yaml v3 made what kustomize considers a breaking change:

```
foo:
  - bar
```

We are leaving the default in tact but adding the option to configure this style so that we can pick up fixes upstream without introducing the breaking change to users. 

Kustomize is currently using its own internal fork to do this, and here is how we configure it in kustomize:

https://github.com/kubernetes-sigs/kustomize/blob/23de1499c27b1d040e90c8094051fe960661196a/kyaml/yaml/alias.go#L55-L59

Reply:
This doesn't look like it's related to the sequence indent issue... I don't see this looking at the configurable `compact_sequence_indent` field, and the newly introduced parameter seems to be hard-coded either to true or false, not be configurable

This looks related to TestNewLinePreserved and doesn't look configurable, is that correct?

Reply:
Oh, my bad! You're right. It came from this commit: https://github.com/go-yaml/yaml/pull/756

Here is the related issue: https://github.com/go-yaml/yaml/issues/755

We ran into this regression in kustomize and needed to fix it.

Reply:
for the patches we're carrying for compatibility reasons with yaml.v2 behavior, can we do the following:
1. Keep each change in its own commit (so in this PR, one commit for the newline issue and commit one for the indenting issue)
2. Add good comments about the scenario being addressed and an explanation of the logic being modified (similar to the request for a comment at https://github.com/kubernetes-sigs/yaml/pull/76#discussion_r951887043). Maybe I'm alone in finding this code difficult to page into working memory, but explaining the mechanics of the tweaks we're making will help our future selves if we have to revisit our change.
3. Add a unit test to the root package demonstrating v2 and our patched v3 behave identically

Reply:
> Maybe I'm alone in finding this code difficult to page into working memory

Absolutely not, this code is _extremely_ confusing and the only reason I hadn't responded to your above comments is because I need to refresh my memory how this code works. 

Thank you for these suggestions, I will update this PR accordingly.

Reply:
> Keep each change in its own commit (so in this PR, one commit for the newline issue and commit one for the indenting issue)

Done

> Add good comments about the scenario being addressed and an explanation of the logic being modified (similar to the request for a comment at Create a fork of go-yaml #76 (comment)). Maybe I'm alone in finding this code difficult to page into working memory, but explaining the mechanics of the tweaks we're making will help our future selves if we have to revisit our change.

I added comments! Hopefully they make it more clear what is happening

> Add a unit test to the root package demonstrating v2 and our patched v3 behave identically

I'm not entirely sure what you're looking for here. `yaml.v2` and `yaml.v3` upstream behave differently (for example, yaml.v2 does the "compact" sequence indent by default while yaml.v3 does not). There are some other marshaling differences between the two as well. 

I added a unit test demonstrating that the upstream yaml.v3 and our patched fork of it behave identically.

Reply:
> > Add a unit test to the root package demonstrating v2 and our patched v3 behave identically
> 
> I'm not entirely sure what you mean by this. `yaml.v2` and `yaml.v3` upstream behave differently (for example, yaml.v2 does the "compact" sequence indent by default while yaml.v3 does not). There are some other subtle marshaling and unmarshalling differences between the two as well.

As we're planning to switch this library's use to our fork of yaml.v3, we would want to make sure we could configure it in a way that preserved yaml.v2 behavior. For each difference we patched yaml.v3 to make consistent with yaml.v2, a test in the root package of this library to demonstrate that patch would be good. Something of this form:

1. setup
2. marshal or unmarshal with yaml.v2
3. marshal or unmarshal with yaml.v3 (maybe with non-default settings)
4. verify result is identical

Reply:
> I added a unit test demonstrating that the upstream yaml.v3 and our patched fork of it behave identically, if that helps.

That wasn't really what I meant, and requires adding a dependency back on upstream yaml.v3, which I don't think we want. I'm more interested in demonstrating consistency between the existing yaml.v2 and our patched yaml.v3

Reply:
Ah, that makes more sense! Apologies for the confusion. I have updated the test accordingly to do what you described above - it verifies that if you use the non-default settings on the yaml.v3 library that I added (compact sequence option, and setting the encoder indent), that yaml.v2 and yaml.v3 behave identically.